File: Baekjoon/sohui/맥주 마시면서 걸어가기.py
def d():
    for i in range(n + 2):
        for j in range(n + 2):
            if i == j:
                continue
            if abs(s[i][0] - s[j][0]) + abs(s[i][1] - s[j][1]) <= 1000:
                s_[i][j] = 1
                s_[j][i] = 1


def dfs(start):
    visit[start] = 1
    for i in range(n + 2):
        if s_[start][i] == 1 and visit[i] == 0:
            dfs(i)


t = int(input())
for i in range(t):
    n = int(input())
    s = [list(map(int, input().split())) for i in range(n + 2)]
    s_ = [[0] * (n + 2) for i in range(n + 2)]
    visit = [0 for i in range(n + 2)]
    d()
    dfs(0)
    if visit[n + 1] == 1:
        print("happy")
    else:
        print("sad")

File: Baekjoon/yerim/맥주 마시면서 걸어가기.cpp
// [9205] 맥주 마시면서 걸어가기
#include <iostream>
#include <queue>
#include <cmath>
#include <cstring>
using namespace std;

bool visited[101];
int store[101][2];
int n, homeX, homeY, festX, festY;

void bfs()
{
    queue<pair<int, int>> q;
    q.push(make_pair(homeX, homeY));
    memset(visited, false, sizeof(visited));
    
    while(!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        if(abs(x - festX) + abs(y - festY) <= 1000) {
            cout << "happy\n";
            return;
        }
        for(int i = 0; i < n; i++) {
            int dist = abs(x - store[i][0]) + abs(y - store[i][1]);
            if(dist <= 1000 && visited[i] == false) {
                q.push(make_pair(store[i][0], store[i][1]));
                visited[i] = true;
            }
        }
    }
    cout << "sad\n";
}

void input()
{
    int t;
    cin >> t;
    while(t--) {
        cin >> n;
        cin >> homeX >> homeY;
        for(int i = 0; i < n; i++) {
            cin >> store[i][0] >> store[i][1];
        }  
        cin >> festX >> festY; 
        bfs();
    }
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    input();

    return 0;
}